### ISSUE TITLE
"The Mainline Kernel"

### ISSUE THREAD

**Issue by ollieparanoid** [postmarketOS/pmbootstrap#91]
One of the goals of pmOS is to unify multiple devices in using one and the same kernel binary.
For a few selected devices, this goal is coming unexpectedly close already - **in the meaning of: the kernel should boot, and we should get output on the serial interface**, allowing further development from there:
* ~~[nokia-rx51 (N900)](https://github.com/postmarketOS/pmbootstrap/wiki/nokia-rx51-%28Nokia-N900%29) ([most hardware is actually working!](http://elinux.org/N900#Kernel_Status))~~
* [lg-hammerhead (Nexus 5)](https://github.com/postmarketOS/pmbootstrap/wiki/lg-hammerhead-%28Google-Nexus-5%29)
* [lg-mako (Nexus 4)](https://github.com/postmarketOS/pmbootstrap/wiki/lg-mako-%28Google-Nexus-4%29)? (assuming it is similar enough to the nexus 7, [ticket](https://github.com/postmarketOS/pmbootstrap/issues/12))
* ["Nexus 7"](https://lwn.net/Articles/680109/) and more devices listed there in the comments ([asus-grouper](https://github.com/postmarketOS/pmbootstrap/wiki/asus-grouper-%28Google-Nexus-7-2012%29) would be supported?)
* [QEMU](https://github.com/postmarketOS/pmbootstrap/pull/56) ;)
* [samsung-apexqtmo (Galaxy S Relay 4G)](https://cascardo.eti.br/blog/News_on_Debian_on_apexqtmo/)

We won't be able to use Alpine's `linux-vanilla` for these devices, because:
* We really need the latest kernel (as support for these devices is being actively developed on)
* We will need custom patches, until they are all mainlined
* We will need different config options for the kernel.

Tasks:
* [x] Create a `linux-postmarketos` aport based on `linux-vanilla`
* [x] Get it working with qemu (should be the easiest)
* [x] Get it working with nokia-rx51
* [x] Get it working with lg-hammerhead
* [x] Close this ticket and continue documentation in the [wiki](https://github.com/postmarketOS/pmbootstrap/wiki/The-mainline-kernel)

**Comment by PabloCastellano** [comment]
JFTR: Sony seems to be a step forward compared to other manufacturers. They have published information on [How to build mainline Linux kernel for Xperia devices](https://developer.sonymobile.com/open-devices/how-to-build-and-flash-a-linux-kernel/how-to-build-mainline-linux-for-xperia-devices/) and [do provide patches upstream](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/log/?qt=author&q=@sonymobile.com).

**Comment by craftyguy** [comment]
the "get it working with nokia-rx51" task is done :)

**Comment by MartijnBraam** [comment]
I've compiled a mainline kernel with all the common smartphones enabled (msm, omap, broadcom, mediatek, alwinner, exynos). Here are the kernel sizes compared:

rx51-only kernel: 3.6 MB
kernel with all common smartphone platforms: 4.3 MB

I've not enabled extra drivers for these platforms, just enabled the platform support itself. Not all extra driver stuff can be build as module so it will be slightly larger than this.

**Comment by ollieparanoid** [comment]
Okay, judging from that and how we must keep an eye on the size (#126), it makes lots of sense to ship one kernel binary per board, as you suggested. (Which does not make them any less mainline).

**Comment by z3ntu** [comment]
Question: How would this work with the different dts files? While developing the mainline kernel for the FP2, I always just appended the correct dtb to the zImage so is there a way of really having one for all? Or is the goal just "the same source for all"?

**Comment by ollieparanoid** [comment]
We would have multiple `dtb` files generated, and put the name of the right `dtb` in the `deviceinfo`.
`postmarketos-mkinitfs` appends the correct one then.
That's how it already works for the N900 and the qemu armhf "device" (look in their source for more information).

**Comment by MartijnBraam** [comment]
@ollieparanoid in the checkboxes above you've added hammerhead, but the hardware for hammerhead is way less supported in mainline than in the lineage kernel and I can add only one kernel package to the hammerhead device file. how should this be handled? A bunch of other devices are gonna have this problem.

An nice overview is from this post on the maemo forums:
https://talk.maemo.org/showpost.php?p=1527669&postcount=1

stuff like the nexus 5x and 6p are listed as "some mainline support" but none of them have graphics yet like the hammerhead.

**Comment by ollieparanoid** [comment]
Whoa, awesome list!
I recommend adding support to our `postmarketos-mkinitfs` package for these devices, and sticking to the Android kernel by default until the mainline kernel is somewhat usable (screen with touch and usb networking are the very basics if you ask me). Then switching to the mainline kernel, but with the option to still install the Android kernel instead. When mainline has all features implemented, we'd move the Android kernel package to `unmaintained`.

All commands working on the kernel/initramfs have a `--flavor` command already, which can be used to choose a kernel.

So if you had the `hammerhead`, and wanted to run it with mainline, you would do: `pmbootstrap install --add linux-postmarketos`, then `pmbootstrap flasher boot --flavor postmarketos`.

**Comment by craftyguy** [comment]
Should the tasks list be expanded to include new 'official' devices?

**Comment by ollieparanoid** [comment]
I recommend that we close this issue after we added support for hammerhead in the mainline kernel (keeping the mainline kernel optional for hammerhead as written above, because the screen does not work yet). Because when we fill this issue up with mainlining every device (especially the ones where we need to start from scratch), it will get messy.

We could put the mainlining progress of each device in the device's wiki page, and once a specific issue comes up, that someone is actually working on, we could make an issue for that.

What would you think about that?

And not directly answering your question, but related to this issue: What I really would like to have is a "mainlining guide" in the wiki, which describes the steps necessary to get the mainline booting on the device (port the `dtb` file over, if there is one, or write it from scratch by looking at the Android kernel source? How does that work?), no matter how much of a rough draft it would be - we could flesh it out when we're actually doing it. At some point it should be possible to have a relatively straight forward guide (assuming we can put up all the work to really mainline a few devices). *Dear reader, if you know anything about that topic, please start such a wiki page :)* (Also if we would see, that there are repetitive tasks involved, we could write some tooling to assist the mainlining.)

**Comment by craftyguy** [comment]
@ollieparanoid yea that makes sense

**Comment by MartijnBraam** [comment]
I've written a quick wiki page about the mainline stuff we have in the master branch currently:
https://github.com/postmarketOS/pmbootstrap/wiki/The-mainline-kernel

**Comment by z3ntu** [comment]
@ollieparanoid It's pretty easy getting mainline = serial output of the boot process plus shell in the initrd. The really complicated part is getting different components of the device to work. A major issue I have currently is that I don't know how I can find out what of the different attributes for the Qualcomm regulators are needed or not. For other parts I was very lucky that some msm8974 devices already have kinda good mainline support (like hammerhead or honami) so I could copy/follow these devices. So yes, it would be great to have a reference for at least Qualcomm devices to kind of know what you are doing.
Also, for the panel driver, there's a guide: https://github.com/freedreno/freedreno/wiki/DSI-Panel-Driver-Porting
And my Linux tree with very limited FP2 support: https://github.com/z3ntu/linux/commits/FP2-cleaner-3

**Comment by ollieparanoid** [comment]
Outstanding, @MartijnBraam! I have expanded the wiki page a bit and added the DSI panel driver link from @z3ntu. Good to hear, that we have another dev with kernel development knowledge around (most of us do not have that skill yet, including myself)!

**Comment by MartijnBraam** [comment]
I found a flaw in your plan to use --flavor to switch kernels, For the hammerhead the deviceinfo is different because I need to specify the dtb and modules from the kernel which is different between mainline and the lineage kernel. I've just patched the device package in the `feature/linux-postmarketos-hammerhead` branch.

I also don't see a better way to handle kernel flavors that have this much differences but maybe any of you do?

**Comment by ollieparanoid** [comment]
Idea: Let's ignore `dtb` and `modules` if none exist for the current kernel (so only if there are 0 `dtb` files and `0` modules). Would that work?

**Comment by montvid** [comment]
I can test a mainline kernel on Nexus 7 2013 if someone gives me the binaries. I am trying to build the linaro build now.

**Comment by ollieparanoid** [comment]
@montvid, the nexus 7 2013 will very likely not work out of the box and no one has customized our `linux-postmarketos` package for it yet. However, I invite you to familiarize yourself with `pmbootstrap` and try it out for yourself. So if you have some time and want to learn something new: Start with the [porting guide](https://wiki.postmarketos.org/wiki/Porting_guide) and a known working Android kernel first, then do the mainline kernel as second step. If you have further questions, please ask in the [chat](https://wiki.postmarketos.org/wiki/Matrix_and_IRC) or open a new issue here (this issue is more about how we package one mainline kernel, not so much about specific devices). Thanks for your interest in postmarketOS!

**Comment by zhuowei** [comment]
I tested the mainline kernel support for the Nexus 6P: I was able to build a kernel that boots, but it crashes somewhere when initializing drivers, and I couldn't get a crash log since I don't have a serial adapter. There's a presentation by Jeremy McNicoll about the Nexus 6P upstreaming effort at https://events.linuxfoundation.org/sites/events/files/slides/JRM_NEXUS_ELC_2017.pdf which says that the 6P should be able to boot to a serial terminal, so not sure what I'm doing wrong. I'll update the Nexus 6P wiki page with more information later.
